Remove CiliumNode deletion logic from CiliumNode watcher and guarantee CiliumNode's OwnerReference is always set

christarazi opened the pull request:
First commit:

> We don't need to implement this logic for two reasons:
> 
> 1) We rely on CiliumNode resources to be deleted / cleaned up by
>    attaching the corresponding K8s Node as an `ownerReference` in the
>    CiliumNode.
> 2) It is redundant to delete the CiliumNode in response to an
>    event...of the CiliumNode deletion itself.
> 
> In very rare cases, this logic can actually delete a newly created
> CiliumNode by accident (see example below). Instead, keep all deletion
> logic besides the actual K8s API calls (DELETE) and perform a Get() to
> ensure that it's been deleted. Otherwise, log to the user that the
> resource may still exist.
> 
> Example:
> Say an existing node was deleted and then recreated in
> quick succession with the same name. When the node is recreated, the
> agent will be scheduled on it. During bootstrap it'll create a
> corresponding CiliumNode resource. Given that only one Operator is
> operational at any time in a cluster, it is already running on another
> node in the cluster. The node-delete event will first delete the K8s
> node and then trigger a CN-delete via reason 1 from above. It is
> possible for the CN-delete event to be delayed such that it is received
> after the node-create event (the recreate). When the CN-delete event is
> received by the already-running Operator, the CiliumNode watcher logic
> will then trigger (erroneously) another CN-delete, thereby deleting the
> CiliumNode resource while the K8s node is still alive.

Second commit:

> It is impossible to set the OwnerReference if we fail to fetch the
> corresponding Kubernetes Node and the existing CiliumNode resource
> doesn't already have it set. We can rely the OwnerReference to be set
> because this logic was added in v1.6, which is sufficiently earlier
> version of Cilium. [1]
> 
> The reason for doing this is to ensure that the OwnerReference can
> always be set. If we cannot, this should be treated as an error and we
> shouldn't proceed. Cilium should not run in an environment where the
> Kubernetes Node resource is missing.
> 
> [1]: 5c365f2 ("ipam: Automatically
> create CiliumNode resource on startup")

christarazi committed:
ipam, operator: Remove CiliumNode deletion logic from CiliumNode watcher

We don't need to implement this logic for two reasons:

1) We rely on CiliumNode resources to be deleted / cleaned up by
   attaching the corresponding K8s Node as an `ownerReference` in the
   CiliumNode.
2) It is redundant to delete the CiliumNode in response to an
   event...of the CiliumNode deletion itself.

In very rare cases, this logic can actually delete a newly created
CiliumNode by accident (see example below). Instead, keep all deletion
logic besides the actual K8s API calls (DELETE) and perform a Get() to
ensure that it's been deleted. Otherwise, log to the user that the
resource may still exist.

Example:
Say an existing node was deleted and then recreated in
quick succession with the same name. When the node is recreated, the
agent will be scheduled on it. During bootstrap it'll create a
corresponding CiliumNode resource. Given that only one Operator is
operational at any time in a cluster, it is already running on another
node in the cluster. The node-delete event will first delete the K8s
node and then trigger a CN-delete via reason 1 from above. It is
possible for the CN-delete event to be delayed such that it is received
after the node-create event (the recreate). When the CN-delete event is
received by the already-running Operator, the CiliumNode watcher logic
will then trigger (erroneously) another CN-delete, thereby deleting the
CiliumNode resource while the K8s node is still alive.

Fixes: 6d44f4cd ("operator: sync cilium nodes to kvstore instead of k8s nodes")
Signed-off-by: Chris Tarazi <chris@isovalent.com>

pchaigno commented on pkg/ipam/allocator/podcidr/podcidr.go:
```diff
@@ -290,10 +290,8 @@ func syncToK8s(nodeGetterUpdater ipam.CiliumNodeGetterUpdater, ciliumNodesToK8s
-			err = nodeGetterUpdater.Delete(nodeName)
-			if k8sErrors.IsNotFound(err) || k8sErrors.IsGone(err) {
-				err = nil
-			}
```
We still have a case where we rely on this code at https://github.com/cilium/cilium/blob/248831f3027f5d9fb26a5ce85f67b6789171a055/pkg/ipam/allocator/podcidr/podcidr.go#L406-L409

If I understand correctly, we should remove that code as well because we are now only deleting nodes from `nodeManager` in reaction to `CiliumNode` delete events, which are triggered by `Node` delete events. So no need to mark the node to deleted in k8s when we remove from `nodeManager`. Does that make sense?

Do we basically don't need `k8sOpDelete` at all anymore?

christarazi replied:
Copying my response from Slack:

> Looking at the deletion code, it makes sense to keep most of it for internal bookkeeping and etc, but I can't understand the motivation for sending a DELETE to the apiserver. The only thing that's plausible is that it could be a "backup" mechanism in case k8s failed to delete the CN itself and Andre probably thought it would be "harmless".

I'll push a revision with only the DELETE removed.

christarazi committed:
pkg/nodediscovery: Error if impossible to set OwnerReference on CiliumNode

It is impossible to set the OwnerReference if we fail to fetch the
corresponding Kubernetes Node and the existing CiliumNode resource
doesn't already have it set. We can rely the OwnerReference to be set
because this logic was added in v1.6, which is sufficiently earlier
version of Cilium. [1]

The reason for doing this is to ensure that the OwnerReference can
always be set. If we cannot, this should be treated as an error and we
shouldn't proceed. Cilium should not run in an environment where the
Kubernetes Node resource is missing.

[1]: 5c365f2c6d7930dcda0b8f0d5e6b826a64022a4f ("ipam: Automatically
create CiliumNode resource on startup")

Signed-off-by: Chris Tarazi <chris@isovalent.com>

pchaigno commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -347,35 +348,46 @@ func (n *NodeDiscovery) mutateNodeResource(nodeResource *ciliumv2.CiliumNode) er
+		nodeTypes.GetName(),
+	)
+	if err != nil && len(nodeResource.ObjectMeta.OwnerReferences) == 0 {
+		return fmt.Errorf("Kubernetes Node resource is unavailable, setting OwnerReference on CiliumNode is impossible: %w", err)
```
The calling function will retry this 10 times before fatalling: https://github.com/cilium/cilium/blob/a8e3fa283cf40b330035265bf83d63800b296bcf/pkg/nodediscovery/nodediscovery.go#L307-L310

Should we just fatal here or do we expect it could not work on the first try for some reason?

christarazi replied:
Good point. We can fatal if the error is specifically NotFound from the apiserver. Otherwise, retry. Added

aanm commented on operator/cilium_node.go:
```diff
@@ -11,7 +11,6 @@ import (
 	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
```
Can you add a `Fixes: 6d44f4cde0f2 ("operator: sync cilium nodes to kvstore instead of k8s nodes")`  in this commit?

aanm reviewed (approved):
LGTM, only a small nit.
